Ensure root scope types have no namespace set

liggitt opened the pull request:
We need to be stricter in ensuring that no namespace is set for root-scoped types. Otherwise, someone could make it through policy checks under a namespace at one layer, then get escalated when another layer ignores the namespace.

These three places were mishandling root-scoped objects:
1. GetAPIRequestInfo sets the request namespace to "default"
2. BeforeCreate/BeforeUpdate sets ObjectMeta.Namespace to "" but ignores a namespace set in the context
3. The etcd key-building functions for root-scoped objects ignore any namespace in the passed context rather than erroring

This PR:
- [x] Adds error return value to root key function to allow indicating errors when the context has an invalid namespace for the type
- [x] Returns errors early in BeforeCreate/BeforeUpdate when an invalid namespace is set in the context or in the object for an unnamespaced type
- [x] Fixes rest tests to use generated names in minion and namespace tests (were getting AlreadyExists errors which was preventing testing the namespace validation conditions)
- [x] Fixes rest test error checking
- [x] Makes minion and node tests use `ClusterScope()`
- [x] Changes GetAPIRequestInfo to stop defaulting `requestInfo.Namespace` to `default` for root-scoped types

vmarmol commented:
Assigning to @lavalamp, feel free to reassign.

derekwaynecarr commented:
I can review this in the morning to save @lavalamp time.

liggitt commented:
@derekwaynecarr I need to look at the integration tests before you take a look

liggitt committed:
Ensure root scope types have no namespace set

# *** ERROR: *** docs are out of sync between cli and markdown
# run hack/run-gendocs.sh > docs/kubectl.md to regenerate

#
# Your commit will be aborted unless you regenerate docs.
    COMMIT_BLOCKED_ON_GENDOCS

liggitt commented:
@derekwaynecarr rebased and updated, PTAL

derekwaynecarr commented:
I will review in morning

Sent from my iPhone

> On Apr 13, 2015, at 4:26 PM, Jordan Liggitt <notifications@github.com> wrote:
> 
> @derekwaynecarr rebased and updated, PTAL
> 
> —
> Reply to this email directly or view it on GitHub.
>

derekwaynecarr commented on pkg/api/rest/create.go:
```diff
@@ -57,7 +57,14 @@ func BeforeCreate(strategy RESTCreateStrategy, ctx api.Context, obj runtime.Obje
 	} else {
-		objectMeta.Namespace = api.NamespaceNone
+		if ns, ok := api.NamespaceFrom(ctx); ok && ns != api.NamespaceNone {
+			return errors.NewBadRequest("namespace is not allowed on this type")
```
is this a situation that could have happened with legacy url format when ?namespace= could have been provided on any resource even if not namespaced?  wondering if this is something you actually saw happen with v1beta3 style urls.  reading through rest of PR to see if I answer my own question.

derekwaynecarr replied:
I think I would change the message to say:

"When creating an object with kind %v, you must have a context with no namespace."

Basically, not sure if type is better than kind, and I imagine this error is not something an average end-user of system will ever find, and is more likely something someone writing a bad unit test would encounter.

lavalamp replied:
Please note that some types (events) can have both namespaced and non-namespaced objects. Does this handle that case?

liggitt replied:
> I think I would change the message to say:

That message was actually copied from validation.go... do we want a different one here?

liggitt replied:
> Please note that some types (events) can have both namespaced and non-namespaced objects.

Doesn't look like the event registry is setting a CreateStrategy or UpdateStrategy, so this isn't used for Events. If Events had been using this previously, though, it would have disallowed some events to be namespaced and some not...

liggitt replied:
> is this a situation that could have happened with legacy url format when ?namespace= could have been provided on any resource even if not namespaced? wondering if this is something you actually saw happen with v1beta3 style urls

Definitely happens with v1beta1 style URLs. Also can happen when a registry is called directly.

derekwaynecarr commented on pkg/api/rest/resttest/resttest.go:
```diff
@@ -57,6 +57,13 @@ func (t *Tester) ClusterScope() *Tester {
+	if t.clusterScope {
+		return api.NewContext()
+	}
+	return api.NewDefaultContext()
```
nit: not sure how much it impacts the test, but I tend to prefer that we not use a default context and instead use one that has a test namespace value assigned.

liggitt replied:
trivial change, updated

derekwaynecarr commented on pkg/apiserver/handlers.go:
```diff
@@ -331,6 +331,9 @@ func (r *APIRequestInfoResolver) GetAPIRequestInfo(req *http.Request) (APIReques
 	}
 
+	// For types that aren't root scoped, we default to the "default" namespace if none is specified for backwards compatibility
+	namespaceUnlessRootScoped := ""
```
When accessing:

/namespaces/foo

I thought the namespace context had to be `foo` to allow for proper policy evaluation.

This appears to change that.  @deads2k - did we not run into this issue, and did we resolve it elsewhere?

liggitt replied:
This is only used if set, and only gets set in the `else` block below... namespaces are a special case and handled in the `if` block, so this function is still behaving as expected.

derekwaynecarr commented on pkg/apiserver/handlers_test.go:
```diff
@@ -151,6 +151,15 @@ func TestGetAPIRequestInfo(t *testing.T) {
 		{"GET", "/namespaces", "list", "", "", "namespaces", "", "Namespace", "", []string{"namespaces"}},
 		{"GET", "/namespaces/other", "get", "", "other", "namespaces", "", "Namespace", "other", []string{"namespaces", "other"}},
```
If this is still passing, then my previous question about /namespaces/foo having a context of foo is answered that for a namespaced resource, the context is foo.  This means that there could be a problem when updating a namespace based on what I read in update.go above.  Need to check if/how that is handled or if I am missing something.

derekwaynecarr replied:
Can you add a test paths for

POST /namespaces
PUT /namespaces/other
PUT /namespaces/other/status
PUT /namespaces/other/finalize

Thanks

liggitt replied:
This test is passing, and I added the first two tests... doesn't look like GetAPIRequestInfo correctly handled namespace subresources, so I fixed that and updated the namespace subresource tests for `status` and `finalize`

Looking at it though, I think the changes in `BeforeUpdate` will cause problems for namespaces... will dig there

liggitt replied:
So, the namespace etcd tests kept passing because I had (incorrectly) updated them to have an empty context. I updated the namespace/etcd_test contexts to match reality:

* create, list, watch calls have an empty namespace in the context
* get, update, delete calls have the current namespace in the context

Take a look at the namespace/etcd changes... I'm not thrilled with them, but I think the registry should be explicit about discarding the namespace from the context by actually doing it, rather than just ignoring it in it's key-building function.

derekwaynecarr commented on pkg/registry/controller/etcd/etcd.go:
```diff
@@ -45,7 +45,7 @@ func NewREST(h tools.EtcdHelper) *REST {
 		// Produces a path that etcd understands, to the root of the resource
 		// by combining the namespace in the context with the given prefix
-		KeyRootFunc: func(ctx api.Context) string {
+		KeyRootFunc: func(ctx api.Context) (string, error) {
```
I get that you want to check this at every layer, and I think at one point before we had generic.etcd the shared KeyRootFunc did throw an error, and @lavalamp removed it in the migration because it was kind of tedious to code against.  Note to self to check how tedious this ends-up being as I continue to review.

derekwaynecarr commented on pkg/registry/generic/etcd/etcd.go:
```diff
@@ -127,6 +131,28 @@ func NamespaceKeyFunc(ctx api.Context, prefix string, name string) (string, erro
 }
 
+// NoNamespaceKeyRootFunc is the default function for constructing etcd paths to root-scoped resource directories.
+func NoNamespaceKeyRootFunc(ctx api.Context, prefix string) (string, error) {
```
So this is the use case that obviously did not exist when we only had namespaced resources in the pre-node all things are namespaced era of the project.  I guess what I don't know is if there is a benefit to pro-actively check for a Namespace on a context if the function completely ignores it.  Was there code that was going direct to the registry layer and not through a client where this caused an impact that warrants changing the signature here versus higher up the stack like you did in rest layer?

derekwaynecarr replied:
Did we have a situation where we had non-namespace objects calling a NamespaceKeyFunc instead of the NoNamespaceKeyFunc?  If so, it seems it like it may be better to protect that a NamespaceKeyFunc is not configured to work with a non-namespaced object during configuration of generic etc object.

derekwaynecarr commented on pkg/registry/minion/etcd/etcd.go:
```diff
@@ -54,11 +54,11 @@ func NewStorage(h tools.EtcdHelper, connection client.ConnectionInfoGetter) (*RE
-		KeyRootFunc: func(ctx api.Context) string {
-			return prefix
+		KeyRootFunc: func(ctx api.Context) (string, error) {
+			return etcdgeneric.NoNamespaceKeyRootFunc(ctx, prefix)
```
I do like a common function for this.

derekwaynecarr commented:
Not a huge fan of the change to EtcdKeyRootFunc to return an error.  Feel like that may be overzealous, but I understand your desire to protect all layers.  More concerned about the namespace context when dealing with the namespace resource itself and if it impacted policy.  We can catch up tomorrow to discuss those concerns.

liggitt commented:
If EtcdKeyRootFunc wasn't used to list, I'd be less concerned.

liggitt committed:
Update namespace registry to handle namespaced contexts

# *** ERROR: *** docs are out of sync between cli and markdown
# run hack/run-gendocs.sh > docs/kubectl.md to regenerate

#
# Your commit will be aborted unless you regenerate docs.
    COMMIT_BLOCKED_ON_GENDOCS

# *** ERROR: *** docs are out of sync between cli and markdown
# run hack/run-gendocs.sh > docs/kubectl.md to regenerate

#
# Your commit will be aborted unless you regenerate docs.
    COMMIT_BLOCKED_ON_GENDOCS

liggitt committed:
Use test namespace

# *** ERROR: *** docs are out of sync between cli and markdown
# run hack/run-gendocs.sh > docs/kubectl.md to regenerate

#
# Your commit will be aborted unless you regenerate docs.
    COMMIT_BLOCKED_ON_GENDOCS

# *** ERROR: *** docs are out of sync between cli and markdown
# run hack/run-gendocs.sh > docs/kubectl.md to regenerate

#
# Your commit will be aborted unless you regenerate docs.
    COMMIT_BLOCKED_ON_GENDOCS

liggitt committed:
Add namespace path parsing tests

# *** ERROR: *** docs are out of sync between cli and markdown
# run hack/run-gendocs.sh > docs/kubectl.md to regenerate

#
# Your commit will be aborted unless you regenerate docs.
    COMMIT_BLOCKED_ON_GENDOCS

derekwaynecarr commented:
What do we want to do with this PR?

liggitt commented:
What indeed. Once pre-v1beta3-style urls go away, I'll be a lot more comfortable without this level of checking. Do you know when we plan to remove v1beta1/2?
